### PR TITLE
[2018-02] [metadata] split IL generation code into seperate compilation units.

### PR DESCRIPTION
**Pull request by lewurm**
By adding callbacks for IL generating runtime code, we can link in code
later into the runtime that does such for configurations which do not
need IL generation initially. This is the case if you configure the
runtime with `--enable-minimal=jit,interpreter`.

In such configuration, the build will produce an additional library,
called `libmono-ilgen`.

An embedder can choose to link this library to the final binary and
initialize IL generation with the following three, newly added, API
calls:

* `void mono_marshal_ilgen_init (void)`
* `void mono_method_builder_ilgen_init (void)`
* `void mono_sgen_mono_ilgen_init (void)`

This change is mostly moving code around, but there are minor changes
like introducing the enum `MarshalTypeCheckPositions`, so we avoid
copy/paste constants across compilation units.

This PR is against `2018-02`, because it was developed and tested against it. Relevant XI changes look like this: https://github.com/xamarin/xamarin-macios/pull/3058/commits/d4ac3078568ec06452fe44fa49a4095afaaea45a

https://github.com/mono/mono/issues/6918

### CODE REVIEW

**Comment by vargaz** [comment]
I still think we should not do this, it splits the code which logically belongs together into 2 files, just to save 60k, which could be saved with much less effort by making the debugger loadable.

**Comment by lewurm** [comment]
@vargaz the separation makes sense imho, as it puts code that solely generates IL into another file. Before `marshal.c` was ~12k lines, and after this change `marshal.c` is ~6k and `marshal-ilgen.c` is ~6k too. While there are some quirks (which can be ironed out later), it's mostly a clear separation.

That said, we should still do the debugger change you suggested 🙂

**Comment by vargaz** [comment]
It splits the code which logically belongs together into two parts.

File: mono/metadata/method-builder-ilgen-internals.h [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+#include <mono/metadata/object-internals.h>
+#include <mono/metadata/class-internals.h>
+#include <mono/metadata/opcodes.h>
+#include <mono/metadata/reflection.h>
+#include <mono/metadata/method-builder.h>
+
+/* ilgen version */
+struct _MonoMethodBuilder {
+	MonoMethod *method;
+	char *name;
```
Comment (luhenry) [inline mono/metadata/method-builder-ilgen-internals.h]: You could fix up the types to be eglib types: `char` -> `gchar`, `int` -> `gint`, `unsigned char` -> `guchar`, etc.

File: mono/metadata/marshal.h [hunk 2]
```diff
@@ -247,13 +247,78 @@ typedef struct {
...
+	STELEMREF_OBJECT, /*no check at all*/
+	STELEMREF_SEALED_CLASS, /*check vtable->klass->element_type */
+	STELEMREF_CLASS, /*only the klass->parents check*/
+	STELEMREF_CLASS_SMALL_IDEPTH, /* like STELEMREF_CLASS bit without the idepth check */
+	STELEMREF_INTERFACE, /*interfaces without variant generic arguments. */
+	STELEMREF_COMPLEX, /*arrays, MBR or types with variant generic args - go straight to icalls*/
+	STELEMREF_KIND_COUNT
+};
+
+static const char *strelemref_wrapper_name[] = {
```
Comment (kumpera) [inline mono/metadata/marshal.h]: Do we need this random constant on the header?
We should wrap it under a `mono_marshal_get_strelemref_wrapper_name`

File: mono/metadata/method-builder.h [hunk 3]
```diff
@@ -21,22 +21,16 @@
...
-	int num_clauses;
-	MonoExceptionClause *clauses;
-	const char **param_names;
-#endif
-} MonoMethodBuilder;
+typedef struct _MonoMethodBuilder MonoMethodBuilder;
+
+#define MONO_METHOD_BUILDER_CALLBACKS_VERSION 1
+
+typedef struct {
```
Comment (kumpera) [inline mono/metadata/method-builder.h]: This looks like the wrong way to split MethodBuilder.

What if we spilt it along the lines of method header builder and method body builder?

This way we can link the whole body builder part into the extra lib.

Comment (lewurm) [reply mono/metadata/method-builder.h]: the body builders already live in `method-builder-ilgen.c`, but aren't actually exposed via the callback API. they are only needed by stuff that lives "behind the ILGEN wall".

does that make sense?

**Comment by lewurm** [comment]
Since this is a rather large change, let me propose the following strategy:
1. if there isn’t something fundamentally wrong with this PR, let’s merge it.
2. backport to master
3. Improve in master, e.g. stuff like changing types, reduce warnings, add/remove callbacks (we have a callback version number 🙂  ) etc.

The most critical stuff that we have to get right in `1.` are the new API functions.

How does that sound to you?

File: mono/metadata/method-builder-ilgen-internals.h (hunk 1 above)
Comment (lewurm) [reply mono/metadata/method-builder-ilgen-internals.h]: good point, let me do that in a follow up commit in master.

File: mono/metadata/marshal.h (hunk 2 above)
Comment (lewurm) [reply mono/metadata/marshal.h]: good point, let me do that in a follow up commit in master.

**Pull request merged**